README
#/* Copyright 1988,1990,1993 by Paul Vixie
# * All rights reserved
# *
# * Distribute freely, except: don't remove my name from the source or
# * documentation (don't take credit for my work), mark your changes (don't
# * get me blamed for your possible bugs), don't alter or remove this
# * notice.  May be sold if buildable source is provided to buyer.  No
# * warrantee of any kind, express or implied, is included with this
# * software; use at your own risk, responsibility for damages (if any) to
# * anyone resulting from the use of this software rests entirely with the
# * user.
# *
# * Send bug reports, bug fixes, enhancements, requests, flames, etc., and
# * I'll try to keep a version up to date.  I can be reached as follows:
# * Paul Vixie          <paul@vix.com>          uunet!decwrl!vixie!paul
# *
# * | Modified by Alexey (allexpert.no) 02.2012
# * | Added jitter support (like in FreeBSD cron)
# * | Added explicit job start delay in seconds; modified crontab looks like:
# * | +<0-59> mm hh dom mon dow command
# * | 
# */

Vixie Cron V3.0
December 27, 1993
[V2.2 was some time in 1992]
[V2.1 was May 29, 1991]
[V2.0 was July 5, 1990]
[V2.0-beta was December 9, 1988]
[V1.0 was May 6, 1987]
Paul Vixie

[Note from Debian cron maintainer: This is the original README from
the the vixie-cron package. The location of many cron files has been
changed in order to comply with Debian policy and common sense -- look
in the cron(8), crontab(1) and crontab(5) man pages for more info, as
well as the README.Debian file in this directory.]

This is a version of 'cron' that is known to run on BSD 4.[23] systems.  It
is functionally based on the SysV cron, which means that each user can have
their own crontab file (all crontab files are stored in a read-protected
directory, usually /var/cron/tabs).  No direct support is provided for
'at'; you can continue to run 'atrun' from the crontab as you have been
doing.  If you don't have atrun (i.e., System V) you are in trouble.

A messages is logged each time a command is executed; also, the files
"allow" and "deny" in /var/cron can be used to control access to the
"crontab" command (which installs crontabs).  It hasn't been tested on
SysV, although some effort has gone into making the port an easy one.

This is more or less the copyright that USENET contributed software usually
has.  Since ATT couldn't use this version if they had to freely distribute
source, and since I'd love to see them use it, I'll offer some rediculously
low license fee just to have them take it.  In the unlikely event that they
do this, I will continue to support and distribute the pseudo-PD version, so
please, don't flame me for wanting my work to see a wider distribution.

To use this: Sorry, folks, there is no cutesy 'Configure' script.  You'll
have to go edit a couple of files... So, here's the checklist:

	Read all the FEATURES, INSTALL, and CONVERSION files
	Edit config.h
	Edit Makefile
		(both of these files have instructions inside; note that
		 some things in config.h are definable in Makefile and are
		 therefore surrounded by #ifndef...#endif)
	'make'
	'su' and 'make install'
		(you may have to install the man pages by hand)
	kill your existing cron process
		(actually you can run your existing cron if you want, but why?)
	build new crontabs using /usr/lib/{crontab,crontab.local}
		(either put them all in "root"'s crontab, or divide it up
		 and rip out all the 'su' commands, collapse the lengthy
		 lists into ranges with steps -- basically, this step is
		 as much work as you want to make it)
	start up the new cron
		(must be done as root)
	watch it. test it with 'crontab -r' and watch the daemon track your
		changes.
	if you like it, change your /etc/{rc,rc.local} to use it instead of
		the old one.

$Id: README,v 2.3 1993/12/28 08:34:43 vixie Exp $


debian/changelog
cron (3.0pl1-116ubuntu3) oneiric; urgency=low

  * Cherry-pick fix from Debian: database.c, cron.c, cron.h,
    debian/copyright:
    - Check orphaned crontabs for adoption. Fix taken from Fedora cronie.
      Closes: #634926, LP: #27520.

 -- Steve Langasek <steve.langasek@ubuntu.com>  Mon, 19 Sep 2011 10:21:01 -0700

cron (3.0pl1-116ubuntu2) oneiric; urgency=low

  * No-change rebuild with current debhelper, to fix maintainer script
    which keeps trying to start cron in my chroots due to bypassing
    invoke-rc.d

 -- Steve Langasek <steve.langasek@ubuntu.com>  Mon, 15 Aug 2011 16:30:24 -0700

cron (3.0pl1-116ubuntu1) natty; urgency=low

  * Merge from debian unstable (LP: #696953), remaining changes:
    - debian/control:
      + Requires debhelper >= 7.3.15ubuntu2 (for Upstart).
      + Move MTA,lockfile-progs to Suggests field.
    - debian/cron.upstart: Add Upstart script.
    - debian/{prerm,postinst,postrm}:
      + Don't call update-rc.d,invoke-rc.d and
        /etc/init.d/cron.
    - debian/rules: Call dh_installinit to install Upstart job properly.

 -- Lorenzo De Liso <blackz@ubuntu.com>  Mon, 03 Jan 2011 20:32:01 +0100

cron (3.0pl1-116) unstable; urgency=high

  * Upload with approval from Release Team to get RC bug fixes in Squeeze
    (see http://lists.debian.org/debian-release/2010/12/msg00719.html)
  * do_command.c, popen.c:
    - Use fork() instead of vfork().
  * do_command.c:
    - Close an unused stream in the fork()ed child prior to exec'ing the
      user's command, thereby avoiding an fd leak. Closes: #604181, LP: #665912
      Previously to this, in conjunction with LVM, the fd leak may have the
      effect of the user being spammed by warnings every time a cron job was
      executed.
  * crontab.5:
    - Fixed the example demonstrating how to run a job on a certain weekday of
      the month (date range was off-by-one). Also, the same example contained
      a superfluous escape, resulting in wrong output. Closes: #606325
  * cron.init:
    - Added $named to Should-Start, in case @reboot jobs need DNS resolution.
      Closes: #602903
    - Added nslcd to Should-Start. LP: #27520 

 -- Christian Kastner <debian@kvr.at>  Wed, 10 Nov 2010 21:13:19 +0100

cron (3.0pl1-115ubuntu1) natty; urgency=low

  * Merge from Debian testing, Ubuntu remaining changes:
    - debian/control:
      + Requires debhelper >= 7.3.15ubuntu2 (for Upstart).
      + Move MTA,lockfile-progs to Suggests field.
    - debian/cron.upstart: Add Upstart script.
    - debian/{prerm,postinst,postrm}:
      + Don't call update-rc.d,invoke-rc.d and
        /etc/init.d/cron.
    - debian/rules:
      + Call dh_installinit to install Upstart job properly.

 -- Alessio Treglia <quadrispro@ubuntu.com>  Sat, 27 Nov 2010 10:46:19 +0100

cron (3.0pl1-115) unstable; urgency=high

  * cron.c: set LC_COLLATE to 'C' in order to properly work when locales
    are set that change the collation order expected by the regular
    expressions used in load_database(). By setting locale we were
    skipping some crontabs unexpectedly due to [a-z] not matching all the ascii
    characters in some languages (such as Estonian).  (Closes: #600310)
    Thanks to Michael Tokarev for spotting this issue and for
    Bastian Blank for pointing for the fix.
  * debian/cron.init: 
      - Fix typo that prevented it from properly loading LC_CTYPE (Closes:
        #600082)
      - Force LC_COLLATE to 'C' regardless of user locale settings.

 -- Javier Fernandez-Sanguino Pen~a <jfs@debian.org>  Sat, 16 Oct 2010 13:05:02 +0200

cron (3.0pl1-114ubuntu1) maverick; urgency=low

  * Merge from debian unstable for documentation updates and reading of
    /etc/default/locale. Remaining changes:
    - debian/control:
      + Build-Depends on debhelper >= 7.3.15ubuntu2, for Upstart
      + Drop MTA and lockfile-args to Suggests
    - add debian/cron.upstart
    - debian/postinst: remove calls to update-rc.d, invoke-rc.d and
      /etc/init.d/cron
    - debian/postrm: remove call to update-rc.d
    - debian/prerm: remove calls to invoke-rc.d and /etc/init.d/cron
    - debian/rules: install Upstart job

 -- Jamie Strandboge <jamie@ubuntu.com>  Tue, 24 Aug 2010 14:56:42 -0500

cron (3.0pl1-114) unstable; urgency=low

  * Cleanup of the package in preparation of the next Debian release.
     - Notice that only one change introduces new behaviour (debian/cron.pam),
       the other changes are just documentation fixes or cleanups to the
       packaging.
  * debian/cron.pam:
    - In addition to the reading /etc/security/pam_env.conf and 
      /etc/environment (obsolete config file) previous default,
      cron will now also read /etc/default/locale in order to setup
      the proper localisation environment for tasks.
  [ Documentation fixes ]
  * debian/cron.{default,init}, cron.8, crontab.5:
    - Updated documentation (comments in scripts and manpages) regarding the
      reading and setting of locale information for the cron daemon and for
      cron tasks to document the new behaviour due to the above change.
  * debian/{preinst,postinst,postrm}:
    - Since /etc/cron.monthly/standard does not provide any worthwhile task
      anymore, remove obsolete /etc/cron.monthly/standard after upgrading
      (local changes are preserved, however). Closes: #585680
  * crontab.1
    - Minor typo fixes to the manpage.
  * crontab.5:
    - Added an example submitted by jidanni (thanks!), showing how to run a
      command on the second Saturday of the month. Closes: #584514 
    - Minor fixes
  [ Fixes to the packaging ]
  * debian/control:
    - Bumped Standards-Version to 3.9.1 (no changes needed)
    - Added Pre-Depends for dpkg (>= 1.15.7.2) for a dpkg-maintscript-helper
      with support for safely removing conffiles
    - Added Homepage: field
    - Added Vcs-* fields
    - Upgraded debhelper dependency version for dh_bugscripts
  * debian/rules:
    - Replace deprecated dh_clean -k with dh_prep
  * debian/NEWS:

cron.h
/* Copyright 1988,1990,1993,1994 by Paul Vixie
 * All rights reserved
 *
 * Distribute freely, except: don't remove my name from the source or
 * documentation (don't take credit for my work), mark your changes (don't
 * get me blamed for your possible bugs), don't alter or remove this
 * notice.  May be sold if buildable source is provided to buyer.  No
 * warrantee of any kind, express or implied, is included with this
 * software; use at your own risk, responsibility for damages (if any) to
 * anyone resulting from the use of this software rests entirely with the
 * user.
 *
 * Send bug reports, bug fixes, enhancements, requests, flames, etc., and
 * I'll try to keep a version up to date.  I can be reached as follows:
 * Paul Vixie          <paul@vix.com>          uunet!decwrl!vixie!paul
 */

/* cron.h - header for vixie's cron
 *
 * $Id: cron.h,v 2.10 1994/01/15 20:43:43 vixie Exp $
 *
 * vix 14nov88 [rest of log is in RCS]
 * vix 14jan87 [0 or 7 can be sunday; thanks, mwm@berkeley]
 * vix 30dec86 [written]
 */

/* reorder these #include's at your peril */

#include <sys/types.h>
#include <sys/param.h>
#include "compat.h"

#include <stdio.h>
#include <ctype.h>
#include <bitstring.h>
#include <pwd.h>
#include <sys/wait.h>

#include "pathnames.h"
#include "config.h"
#include "externs.h"

#if SYS_TIME_H
# include <sys/time.h>
#else
# include <time.h>
#endif


#ifdef WITH_SELINUX
#include <selinux/selinux.h>
#endif

#define SYSUSERNAME "root"


	/* these are really immutable, and are
	 *   defined for symbolic convenience only
	 * TRUE, FALSE, and ERR must be distinct
	 * ERR must be < OK.
	 */
#define TRUE		1
#define FALSE		0
	/* system calls return this on success */
#define OK		0
	/*   or this on error */
#define ERR		(-1)

	/* turn this on to get '-x' code */
#ifndef DEBUGGING
#define DEBUGGING	FALSE
#endif

#define READ_PIPE	0	/* which end of a pipe pair do you read? */
#define WRITE_PIPE	1	/*   or write to? */
#define STDIN		0	/* what is stdin's file descriptor? */
#define STDOUT		1	/*   stdout's? */
#define STDERR		2	/*   stderr's? */
#define ERROR_EXIT	1	/* exit() with this will scare the shell */
#define	OK_EXIT		0	/* exit() with this is considered 'normal' */
#define	MAX_FNAME	100	/* max length of internally generated fn */
#define	MAX_COMMAND	1000	/* max length of internally generated cmd */
#define	MAX_TEMPSTR	1000	/* max length of envvar=value\0 strings */
#define	MAX_ENVSTR	MAX_TEMPSTR	/* DO NOT change - buffer overruns otherwise */
#define	MAX_UNAME	20	/* max length of username, should be overkill */
#define	ROOT_UID	0	/* don't change this, it really must be root */
#define	ROOT_USER	"root"	/* ditto */

				/* NOTE: these correspond to DebugFlagNames,
				 *	defined below.
				 */
#define	DEXT		0x0001	/* extend flag for other debug masks */
#define	DSCH		0x0002	/* scheduling debug mask */
#define	DPROC		0x0004	/* process control debug mask */
#define	DPARS		0x0008	/* parsing debug mask */
#define	DLOAD		0x0010	/* database loading debug mask */
#define	DMISC		0x0020	/* misc debug mask */
#define	DTEST		0x0040	/* test mode: don't execute any commands */
#define	DBIT		0x0080	/* bit twiddling shown (long) */

#define	CRON_TAB(u)	"%s/%s", SPOOL_DIR, u
#define	REG		register
#define	PPC_NULL	((char **)NULL)

#ifndef MAXHOSTNAMELEN
#define MAXHOSTNAMELEN 64
#endif

#define	Skip_Blanks(c, f) \
			while (c == '\t' || c == ' ') \
				c = get_char(f);

#define	Skip_Nonblanks(c, f) \
			while (c!='\t' && c!=' ' && c!='\n' && c != EOF) \
				c = get_char(f);

#define	Skip_Line(c, f) \
			do {c = get_char(f);} while (c != '\n' && c != EOF);

#if DEBUGGING
# define Debug(mask, message) \
			if ( (DebugFlags & (mask) )  ) \
				printf message;
#else /* !DEBUGGING */
# define Debug(mask, message) \
			;
#endif /* DEBUGGING */

#define	MkLower(ch)	(isupper(ch) ? tolower(ch) : ch)
#define	MkUpper(ch)	(islower(ch) ? toupper(ch) : ch)
#define	Set_LineNum(ln)	{Debug(DPARS|DEXT,("linenum=%d\n",ln)); \
			 LineNumber = ln; \
			}

typedef int time_min;

#define SECONDS_PER_MINUTE 60

#define	FIRST_MINUTE	0
#define	LAST_MINUTE	59
#define	MINUTE_COUNT	(LAST_MINUTE - FIRST_MINUTE + 1)

#define	FIRST_HOUR	0
#define	LAST_HOUR	23
#define	HOUR_COUNT	(LAST_HOUR - FIRST_HOUR + 1)

#define	FIRST_DOM	1
#define	LAST_DOM	31
#define	DOM_COUNT	(LAST_DOM - FIRST_DOM + 1)

#define	FIRST_MONTH	1
#define	LAST_MONTH	12
#define	MONTH_COUNT	(LAST_MONTH - FIRST_MONTH + 1)

/* note on DOW: 0 and 7 are both Sunday, for compatibility reasons. */
#define	FIRST_DOW	0
#define	LAST_DOW	7
#define	DOW_COUNT	(LAST_DOW - FIRST_DOW + 1)

			/* each user's crontab will be held as a list of
			 * the following structure.
			 *
			 * These are the cron commands.
			 */

typedef	struct _entry {
	struct _entry	*next;
	uid_t		uid;	
	gid_t		gid;
	char		**envp;
	char		*cmd;
	int	        second; /* added by alexey:
						   Amount of seconds to sleep before starting a job;
						   has the same function as -j jitter. Unlike the latter
						   is not random but defined explicitly in crontab per 
						   job as follows:
						   +<seconds> <minute settings> <hour settings> ...*/
	bitstr_t	bit_decl(minute, MINUTE_COUNT);
	bitstr_t	bit_decl(hour,   HOUR_COUNT);
	bitstr_t	bit_decl(dom,    DOM_COUNT);
	bitstr_t	bit_decl(month,  MONTH_COUNT);
	bitstr_t	bit_decl(dow,    DOW_COUNT);
	int		flags;
#define	DOM_STAR	0x01
#define	DOW_STAR	0x02
#define	WHEN_REBOOT	0x04
#define MIN_STAR	0x08
#define HR_STAR		0x10
} entry;

			/* the crontab database will be a list of the
			 * following structure, one element per user
			 * plus one for the system.
			 *
			 * These are the crontabs.
			 */

typedef	struct _user {
	struct _user	*next, *prev;	/* links */
	char		*name;
	time_t		mtime;		/* last modtime of crontab */
	entry		*crontab;	/* this person's crontab */
#ifdef WITH_SELINUX
        security_context_t scontext;    /* SELinux security context */
#endif
} user;

typedef	struct _cron_db {
	user		*head, *tail;	/* links */
	time_t		user_mtime;     /* last modtime on spooldir */
	time_t		sys_mtime;      /* last modtime on system crontab */
#ifdef DEBIAN
	time_t		sysd_mtime;     /* last modtime on system crondir */
#endif
} cron_db;

typedef struct _orphan {
	struct _orphan  *next;          /* link */
	char            *uname;
	char            *fname;
	char            *tabname;
} orphan;

void		set_cron_uid __P((void)),
		set_cron_cwd __P((void)),
		load_database __P((cron_db *)),
		open_logfile __P((void)),
		sigpipe_func __P((void)),
		job_add __P((entry *, user *)),
		do_command __P((entry *, user *)),
		link_user __P((cron_db *, user *)),
		unlink_user __P((cron_db *, user *)),
		free_user __P((user *)),
		env_free __P((char **)),
		unget_char __P((int, FILE *)),
		free_entry __P((entry *)),
		acquire_daemonlock __P((int)),
		skip_comments __P((FILE *)),
		log_it __P((char *, int, char *, char *)),
		log_close __P((void)),
		check_orphans __P((cron_db *));

int		job_runqueue __P((void)),
		set_debug_flags __P((char *)),
		get_char __P((FILE *)),
		get_string __P((char *, int, FILE *, char *)),
		swap_uids __P((void)),
		swap_uids_back __P((void)),
		load_env __P((char *, FILE *)),
		cron_pclose __P((FILE *)),
		strcmp_until __P((char *, char *, int)),
		allowed __P((char *)),
		strdtb __P((char *));

long            get_gmtoff(time_t *, struct tm *);

char		*env_get __P((char *, char **)),
		*arpadate __P((time_t *)),
		*mkprints __P((unsigned char *, unsigned int)),
		*first_word __P((char *, char *)),
		**env_init __P((void)),
		**env_copy __P((char **)),
		**env_set __P((char **, char *));

user		*load_user __P((int, struct passwd *, char *, char *, char *)),
		*find_user __P((cron_db *, char *));

entry		*load_entry __P((FILE *, void (*)(),
				 struct passwd *, char **));

FILE		*cron_popen __P((char *, char *, entry *));


				/* in the C tradition, we only create
				 * variables for the main program, just
				 * extern them elsewhere.
				 */

#ifdef MAIN_PROGRAM
# if !defined(LINT) && !defined(lint)
char	*copyright[] = {
		"@(#) Copyright 1988,1989,1990,1993,1994 by Paul Vixie",
		"@(#) All rights reserved"
	};
# endif

char	*MonthNames[] = {
		"Jan", "Feb", "Mar", "Apr", "May", "Jun",
		"Jul", "Aug", "Sep", "Oct", "Nov", "Dec",
		NULL
	};

char	*DowNames[] = {
		"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun",
		NULL
	};

char	*ProgramName;
int	LineNumber;
unsigned Jitter,
	RootJitter;
time_t	StartTime;
time_min timeRunning;
time_min virtualTime;
time_min clockTime;
static long GMToff;

int	stay_foreground;
int     lsbsysinit_mode;
int     log_level = 1;
char    cron_default_mail_charset[MAX_ENVSTR] = "";

# if DEBUGGING
int	DebugFlags;
char	*DebugFlagNames[] = {	/* sync with #defines */
		"ext", "sch", "proc", "pars", "load", "misc", "test", "bit",
		NULL		/* NULL must be last element */
	};
# endif /* DEBUGGING */
#else /*MAIN_PROGRAM*/
extern	char	*copyright[],
		*MonthNames[],
		*DowNames[],
		*ProgramName;
extern  int     lsbsysinit_mode;
extern  int     log_level;
extern	int	LineNumber;
extern unsigned	Jitter,
		RootJitter;
extern	time_t	StartTime;
extern  time_min timeRunning;
extern  time_min virtualTime;
extern  time_min clockTime;
extern  char     cron_default_mail_charset[MAX_ENVSTR];
# if DEBUGGING
extern	int	DebugFlags;
extern	char	*DebugFlagNames[];
# endif /* DEBUGGING */
#endif /*MAIN_PROGRAM*/
